### ISSUE TITLE
Add compiler warning for double semicolon

### ISSUE THREAD

**Issue by hellow554** [rust-lang/rust#60876]
This could should emit an error:

```
pub fn foo() {
    return;;
}
```

I think something similar to this would be a good one:

```
warning: redundant `;` detected
 --> src/main.rs:X:XX
  |
LL|         return;;
  |                ^
  |                |
  |                help: remove this `;` (this should be a suggestion so rustfix can fix it automatically)
```

**Comment by hbina** [comment]
Hi, I would like to resolve this issue. Do you have any lead/pointers that I can use to start?

**Comment by hellow554** [comment]
I'm not entirely sure, I only started recently with hacking the compiler. I think this can be easily done by looking at 

https://github.com/rust-lang/rust/blob/49d139c64b69ec5289f9f81db885ecfc2c7a8366/src/libsyntax/parse/lexer/mod.rs#L947-L950

You could do something similar to this:

```
';' => { 
    self.bump(); 
    if self.ch_is(';') {
        let mut err = self.sess.span_diagnostic.struct_span_warn(self.sp, "redundant `;` detected");
        ...
        err.emit();
    }
    Ok(token::Semi) 
} 
```

Of course you should also detect more than one, e.g. `return ;;;`. Also make sure it works for `let a = "";;;;;;`.

**Comment by hbina** [comment]
is it safe to just perform a while loop?
i am not sure how reliable this solution could be.
Especially if there's a space like `; ;`

**Comment by hellow554** [comment]
> is it safe to just perform a while loop?

Go and try it. 

>  Especially if there's a space like ; ;

That could be a problem indeed, because spaces are treated seperatly.

> Regardless how to say that this ;;;;;; entirely line is faulty, not just the first character?

You mean something like 

```
fn foo() {
    ;;;;;;;;;;;;;;
}
```

? I think this should be a different warning. Something like *"sole `;` detected"*, because it does not end any statement at all and should have no right to exist.

I think a person which is more privileged than me should say a word or too as well :) I'm not a rust lang member

**Comment by hellow554** [comment]
Instead of using the lexer level, you could also use the token level, because the lexer converts the bytestream to tokens (as you can see with `Ok(token::Semi)` and if you detect two Semi tokens next to each other, you could emit a warning there. Should be also possible. But I'm not entirely sure where to do that.

**Comment by Centril** [comment]
cc @petrochenkov 

This should be done as new a lint pass (`redundant_semicolon`) or included in an existing one.

**Comment by hbina** [comment]
@Centril how does one go about doing that?

**Comment by scottmcm** [comment]
A rustfix-able lint for this (along the line of `unused_parens`, etc) sounds great.  (Not an error, of course.)

**Comment by Centril** [comment]
We discussed this in the language team meeting today. People felt generally positive about such a lint. However, we would like to not lint on `return/break/continue;` and `return/break x;` specifically.

**Comment by varkor** [comment]
@hbina: I can give more detailed mentoring advice later, but for now: take a look at an existing lint (e.g. `UNUSED_MUST_USE`). If you search the compiler source for that, you'll find a bunch of places you need to modify to add a new lint. You'll want to add a new lint type (e.g. `REDUNDANT_SEMICOLON`) and add a new pass that checks for an empty statement — you'll need to work out what this looks like: something like `hir::StmtKind::Semi(..)` (but I'm not completely sure what kind of expression it'll contain, so this will involve some testing) — and emits a lint.

**Comment by gautammohan** [comment]
Is this issue currently being worked on? I'm interested in contributing.

**Comment by nathanwhit** [comment]
I've done some work on this, and have a working draft ready for review. The PR is at #62984, feedback is appreciated!

**Comment by BO41** [comment]
> We discussed this in the language team meeting today. People felt generally positive about such a lint. However, we would like to not lint on `return/break/continue;` and `return/break x;` specifically.

@Centril Would you mind giving a short explanation? I am curious :smiley:

**Comment by Centril** [comment]
@BO41 What sort of explanation? We felt positive about this lint since it removes something which sometimes happens and which is a mistake, iirc. We don't want to lint on `return x;` since that is commonly accepted as OK/good style.

**Comment by iluuu1994** [comment]
@Centril I think this issue is only talking about duplicate semicolons. `return x;;` should warn but `return x;` shouldn't. This also shouldn't:

```rust
fn fn_returning_void() {
    other_fn_returning_void();
}
```

**Comment by scottmcm** [comment]
@BO41 well, `rustfmt` puts semicolons after `return` expressions, so it wouldn't be nice to tell people to remove something only for rustfmt to put it back again 🙂

**Comment by BO41** [comment]
Thanks @Centril and @scottmcm for the explanation.
I think we got a bit confused, as @iluuu1994 described.

This issue is about multiple semicolons. Not a single after `return`.
So _I_ think 
>we would like to not lint on `return/break/continue;` and `return/break x;` specifically.

is not really what this is about. As `return;;` is wrong. And `continue;;;;;;;;;` too.
But I agree suggesting from `break;` to `break` is bad.
